File: python/wenet.py
"""
It requires a python wrapped c++ ctc decoder.
Please install it by following:
https://github.com/Slyne/ctc_decoder.git
"""

from __future__ import print_function

import argparse
import copy
import logging
import os
import sys
sys.dont_write_bytecode = True
sys.path.append(os.getcwd()+"/swig_decoders")

import torch
import yaml
from torch.utils.data import DataLoader

from dataset.dataset import Dataset
from utils.common import IGNORE_ID
from utils.file_utils import read_symbol_table
from utils.file_utils import read_lists

import multiprocessing
import numpy as np

from swig_decoders import map_batch,ctc_beam_search_decoder_batch,TrieVector, PathTrie
from utils.sophon_inference import SophonInference

import contextlib
import wave
import time

import logging
logging.basicConfig(level=logging.INFO)

def ctc_decoding(beam_log_probs, beam_log_probs_idx, encoder_out_lens, vocabulary, mode='ctc_prefix_beam_search'):
    beam_size = beam_log_probs.shape[-1]
    batch_size = beam_log_probs.shape[0]
    num_processes = min(multiprocessing.cpu_count(), batch_size)
    hyps = []
    score_hyps = []
    
    if mode == 'ctc_greedy_search':
        if beam_size != 1:
            log_probs_idx = beam_log_probs_idx[:, :, 0]
        batch_sents = []
        for idx, seq in enumerate(log_probs_idx):
            batch_sents.append(seq[0:encoder_out_lens[idx]].tolist())
        hyps = map_batch(batch_sents, vocabulary, num_processes,
                         True, 0)
    elif mode in ('ctc_prefix_beam_search', "attention_rescoring"):
        batch_log_probs_seq_list = beam_log_probs.tolist()
        batch_log_probs_idx_list = beam_log_probs_idx.tolist()
        batch_len_list = encoder_out_lens.tolist()
        batch_log_probs_seq = []
        batch_log_probs_ids = []
        batch_start = []  # only effective in streaming deployment
        batch_root = TrieVector()
        root_dict = {}
        for i in range(len(batch_len_list)):
            num_sent = batch_len_list[i]
            batch_log_probs_seq.append(
                batch_log_probs_seq_list[i][0:num_sent])
            batch_log_probs_ids.append(
                batch_log_probs_idx_list[i][0:num_sent])
            root_dict[i] = PathTrie()
            batch_root.append(root_dict[i])
            batch_start.append(True)
        score_hyps = ctc_beam_search_decoder_batch(batch_log_probs_seq,
                                                   batch_log_probs_ids,
                                                   batch_root,
                                                   batch_start,
                                                   beam_size,
                                                   num_processes,
                                                   0, -2, 0.99999)
        if mode == 'ctc_prefix_beam_search': 
            for cand_hyps in score_hyps:
                hyps.append(cand_hyps[0][1])
            hyps = map_batch(hyps, vocabulary, num_processes, False, 0)
    return hyps, score_hyps
            
def adjust_feature_length(feats, length, padding_value=0):
    # Adjust the length of the feature to a uniform length
    # feats: B*T*L tensor, where B is the batch size, T is the length of the feature, and L is the size of the feature
    # length: Length of adjusted T
    if feats.shape[1] < length:
        B, T, L = feats.shape
        tmp = np.full([B, length-T, L], padding_value)
        feats = np.concatenate((feats, tmp), axis=1)
    elif feats.shape[1] > length:
        feats = feats[:, :length, :]
    return feats

def calculate_total_time(data_list):
    # Calculate the total duration of all wav data files to measure the inference speed of the model
    lists = read_lists(data_list)
    total_time = 0
    for _, list in enumerate(lists):
        list = eval(list)
        wav_file_path = list["wav"]
        with contextlib.closing(wave.open(wav_file_path, 'r')) as f:
            frames = f.getnframes()
            rate = f.getframerate()
            total_time += frames / float(rate)
    return total_time
    
def get_args():
    parser = argparse.ArgumentParser(description='recognize with your model')
    
    parser.add_argument('--input', default='../datasets/aishell_S0764/aishell_S0764.list', help='path of input')
    parser.add_argument('--bmodel', default='../models/BM1684/wenet_encoder_fp32.bmodel', help='path of bmodel')
    parser.add_argument('--dev_id', type=int, default=0, help='dev id')
    parser.add_argument('--result_file', default='./result.txt', help='asr result file')
    parser.add_argument('--mode',
                        choices=[
                            'ctc_greedy_search', 'ctc_prefix_beam_search'],
                        default='ctc_prefix_beam_search',
                        help='decoding mode')
    parser.add_argument('--data_type',
                        default='raw',
                        choices=['raw', 'shard'],
                        help='train and cv data type')
    parser.add_argument('--dict', default='../config/lang_char.txt', help='dict file')
    parser.add_argument('--config', default='../config/train_u2++_conformer.yaml', help='config file')
    parser.add_argument('--batch_size',
                        type=int,
                        default=1,
                        help='asr result file')
    parser.add_argument('--decoding_chunk_size',
                        default=16,
                        type=int,
                        help='decoding chunk size for dynamic chunk')
    parser.add_argument('--num_decoding_left_chunks',
                        default=5,
                        type=int,
                        help='number of left chunks, this is for decoding, the chunk size is decoding_chunk_size')
    parser.add_argument('--subsampling_rate',
                        default=4,
                        type=int,
                        help='It is related to the subsampling module adopted by the model, see the definition in the subsampling.py file')
    parser.add_argument('--context',
                        default=7,
                        type=int,
                        help='It is related to the subsampling module adopted by the model, see the definition in the subsampling.py file')
    args = parser.parse_args()
    # print(args)
    return args

if __name__ == '__main__':
    args = get_args()
    logging.basicConfig(level=logging.DEBUG,
                        format='%(asctime)s %(levelname)s %(message)s')

    with open(args.config, 'r') as fin:
        configs = yaml.load(fin, Loader=yaml.FullLoader)

    reverse_weight = configs["model_conf"].get("reverse_weight", 0.0)
    symbol_table = read_symbol_table(args.dict)
    test_conf = copy.deepcopy(configs['dataset_conf'])
    test_conf['filter_conf']['max_length'] = 102400
    test_conf['filter_conf']['min_length'] = 0
    test_conf['filter_conf']['token_max_length'] = 102400
    test_conf['filter_conf']['token_min_length'] = 0
    test_conf['filter_conf']['max_output_input_ratio'] = 102400
    test_conf['filter_conf']['min_output_input_ratio'] = 0
    test_conf['speed_perturb'] = False
    test_conf['spec_aug'] = False
    test_conf['spec_sub'] = False
    test_conf['spec_trim'] = False
    test_conf['shuffle'] = False
    test_conf['sort'] = False
    test_conf['fbank_conf']['dither'] = 0.0
    test_conf['batch_conf']['batch_type'] = "static"
    test_conf['batch_conf']['batch_size'] = args.batch_size
    
    start_time = time.time()
    test_dataset = Dataset(args.data_type,
                           args.input,
                           symbol_table,
                           test_conf,
                           bpe_model=None,
                           partition=False)
    preprocess_time = time.time() - start_time

    test_data_loader = DataLoader(test_dataset, batch_size=None, num_workers=0)

    # Init encoder
    encoder = SophonInference(model_path=args.bmodel, device_id=args.dev_id, input_mode=0)

    # Load dict
    vocabulary = []
    char_dict = {}
    with open(args.dict, 'r') as fin:
        for line in fin:
            arr = line.strip().split()
            assert len(arr) == 2
            char_dict[int(arr[1])] = arr[0]
            vocabulary.append(arr[0])
    eos = sos = len(char_dict) - 1
    
    subsampling = args.subsampling_rate
    context = args.context  
    stride = subsampling * args.decoding_chunk_size
    decoding_window = (args.decoding_chunk_size - 1) * subsampling + context        
    required_cache_size = args.decoding_chunk_size * args.num_decoding_left_chunks

    output_size = configs["encoder_conf"]["output_size"]
    num_layers = configs["encoder_conf"]["num_blocks"]
    cnn_module_kernel = configs["encoder_conf"].get("cnn_module_kernel", 1) - 1
    head = configs["encoder_conf"]["attention_heads"]
    d_k = configs["encoder_conf"]["output_size"] // head
    
    inference_time = 0.0
    postprocess_time = 0.0
    # Start speech recognition
    with torch.no_grad(), open(args.result_file, 'w') as fout:
        for _, batch in enumerate(test_data_loader):
            keys, feats, _, feats_lengths, _ = batch
            feats, feats_lengths = feats.numpy(), feats_lengths.numpy()
            
            supplemental_batch_size = args.batch_size - feats.shape[0]
            
            att_cache = np.zeros((args.batch_size, num_layers, head, required_cache_size, d_k * 2), dtype=np.float32)
            cnn_cache = np.zeros((args.batch_size, num_layers, output_size, cnn_module_kernel), dtype=np.float32)
            cache_mask = np.zeros((args.batch_size, 1, required_cache_size), dtype=np.float32)
            offset = np.zeros((args.batch_size, 1), dtype=np.int32)
            
            encoder_out = []
            beam_log_probs = []
            beam_log_probs_idx = []
            
            num_frames = feats.shape[1]
            for cur in range(0, num_frames - context + 1, stride):
                end = min(cur + decoding_window, num_frames)
                chunk_xs = feats[:, cur:end, :]
                if chunk_xs.shape[1] < decoding_window:
                    chunk_xs = adjust_feature_length(chunk_xs, decoding_window, padding_value=0)
                    chunk_xs = chunk_xs.astype(np.float32)
                chunk_lens = np.full(args.batch_size, fill_value=chunk_xs.shape[1], dtype=np.int32)

                encoder_input = [chunk_lens, att_cache, cnn_cache, chunk_xs, cache_mask, offset]
                start_time = time.time()
                out_dict = encoder.infer_numpy(encoder_input)
                inference_time += time.time() - start_time
                
                chunk_out = out_dict["chunk_out"]
                chunk_log_probs = out_dict["log_probs"]
                chunk_log_probs_idx = out_dict["log_probs_idx"].astype(np.int32)
                att_cache = out_dict['r_att_cache']
                cnn_cache = out_dict['r_cnn_cache']
                offset = out_dict['r_offset'].astype(np.int32)
                cache_mask = out_dict['r_cache_mask']
                chunk_out_lens = out_dict['chunk_out_lens'].astype(np.int32)

                encoder_out.append(chunk_out)
                beam_log_probs.append(chunk_log_probs)
                beam_log_probs_idx.append(chunk_log_probs_idx)
                
                # ctc decode
                start_time = time.time()
                chunk_hyps, _ = ctc_decoding(chunk_log_probs, chunk_log_probs_idx, chunk_out_lens, vocabulary)
                postprocess_time += time.time() - start_time
                print(chunk_hyps)
           
            encoder_out = np.concatenate(encoder_out, axis=1)
            encoder_out_lens = np.full(args.batch_size, fill_value=encoder_out.shape[1], dtype=np.int32)
            beam_log_probs = np.concatenate(beam_log_probs, axis=1)
            beam_log_probs_idx = np.concatenate(beam_log_probs_idx, axis=1)

            start_time = time.time()
            hyps, score_hyps = ctc_decoding(beam_log_probs, beam_log_probs_idx, encoder_out_lens, vocabulary, args.mode)
            postprocess_time += time.time() - start_time
            
            for i, key in enumerate(keys):
                content = hyps[i]
                logging.info('{} {}'.format(key, content))
                fout.write('{} {}\n'.format(key, content))
                
    logging.info("------------------ Predict Time Info ----------------------")
    total_data_time = calculate_total_time(args.input)
    logging.info("preprocess_time(ms): {:.4f}".format((preprocess_time / total_data_time) * 1000))
    logging.info("inference_time(ms): {:.4f}".format((inference_time / total_data_time) * 1000))
    logging.info("postprocess_time(ms): {:.4f}".format((postprocess_time / total_data_time) * 1000))

File: cpp/src/wenet.cpp
#include "wenet.h"

using namespace bmruntime;

int WeNet::Init(const std::vector<std::string>& dict, int sample_frequency, int num_mel_bins, int frame_shift, int frame_length, int decoding_chunk_size, int subsampling_rate, int context) {
    this->dict = dict;
    this->sample_frequency = sample_frequency;
    this->sample_frequency = sample_frequency;
    this->num_mel_bins = num_mel_bins;
    this->frame_shift = frame_shift;
    this->frame_length = frame_length;
    this->decoding_chunk_size = decoding_chunk_size;
    this->subsampling_rate = subsampling_rate;
    this->context = context;

    // create Network
    std::vector<const char *> network_names;
    ctx->get_network_names(&network_names);
    net = std::make_shared<Network>(*ctx, network_names[0], 0); // use stage[0]
    assert(net->info()->input_num == 6);

    // Initialize the memory space required for the input and output tensors
    inputs = net->Inputs();
    outputs = net->Outputs();
    const bm_tensor_t * log_probs_bm_tensor_t = outputs[0]->tensor();
    batch_size = log_probs_bm_tensor_t->shape.dims[0];
    beam_size = log_probs_bm_tensor_t->shape.dims[2];
    assert(batch_size == 1 && "Streaming inference only supports batch size = 1!");

    return 0;
}

int WeNet::pre_process(const char* file_path) {
    LOG_TS(m_ts, "wenet load file");
    SF_INFO sfinfo;
    SNDFILE* sndfile = sf_open(file_path, SFM_READ, &sfinfo);

    if (sndfile == NULL) {
        std::cerr << "Error opening WAV file: " << sf_strerror(sndfile) << std::endl;
        return 1;
    }

    // std::cout << "Sample rate: " << sfinfo.samplerate << std::endl;
    // std::cout << "Channels: " << sfinfo.channels << std::endl;
    // std::cout << "Frames: " << sfinfo.frames << std::endl;

    // Read the samples into a buffer
    const int buffer_size = sfinfo.frames * sfinfo.channels;
    double buffer[buffer_size];
    sf_read_double(sndfile, buffer, buffer_size);
    // sf_count_t samples_read = sf_read_double(sndfile, buffer, buffer_size);
    // std::cout << "Samples read: " << samples_read << std::endl;

    // Close the file
    sf_close(sndfile);

    arma::fmat waveform = arma::fmat(sfinfo.channels, sfinfo.frames);
    if(sfinfo.channels == 1) {
        for (int i = 0; i < sfinfo.frames; i++) {
            waveform(0, i) = buffer[i] * (1 << 15);
        }
    }
    else if(sfinfo.channels == 2) {
        for (int i = 0; i < sfinfo.frames; i += 2) {
            waveform(0, i / 2) = buffer[i] * (1 << 15);
            waveform(0, i / 2) = buffer[i + 1] * (1 << 15);
        }       
    }
    else {
        std::cerr << "The number of channels in the wav file is not normal!" << std::endl;
        return 1;
    }
    LOG_TS(m_ts, "wenet load file");

    LOG_TS(m_ts, "wenet fbank");
    feats = fbank(waveform, num_mel_bins, frame_length, frame_shift, sample_frequency, 0.0, 0.0, true, true, false);
    // todo:spec_sub
    LOG_TS(m_ts, "wenet fbank");

    return 0;
}

int WeNet::inference() {
    void* att_cache = calloc(inputs[1]->num_elements(), sizeof(float));
    void* cnn_cache = calloc(inputs[2]->num_elements(), sizeof(float));
    void* cache_mask = calloc(inputs[4]->num_elements(), sizeof(float));
    void* offset = calloc(inputs[5]->num_elements(), sizeof(int));

    void* log_probs = calloc(outputs[0]->num_elements(), sizeof(float));
    void* log_probs_idx = calloc(outputs[1]->num_elements(), sizeof(int));
    void* chunk_out = calloc(outputs[2]->num_elements(), sizeof(float));
    void* chunk_out_lens = calloc(outputs[3]->num_elements(), sizeof(int));

    // inference
    int num_frames = feats.n_rows;
    int stride = subsampling_rate * decoding_chunk_size;
    int decoding_window = (decoding_chunk_size - 1) * subsampling_rate + context;
    
    result = "";
    for(int cur = 0; cur < num_frames - context + 1; cur += stride) {
        int end = std::min(cur + decoding_window, num_frames);
        arma::fmat chunk_xs = feats.submat(cur, 0, end - 1, feats.n_cols - 1);
        // pad if needed
        if((int)chunk_xs.n_rows < decoding_window) {
            arma::fmat pad_zeros(decoding_window - chunk_xs.n_rows, chunk_xs.n_cols, arma::fill::zeros);
            chunk_xs = arma::join_cols(chunk_xs, pad_zeros);
        }
        
        void* chunk_xs_ptr = fmat_to_sys_mem(chunk_xs);
        int chunk_lens = chunk_xs.n_rows;
        int* chunk_lens_ptr = (int*) malloc(sizeof(int));
        if (chunk_lens_ptr != nullptr) {
            *chunk_lens_ptr = chunk_lens;
        }
        else {
            std::cerr << "Failed to request memory space" << std::endl;
            exit(1);
        }
        void* void_chunk_lens_ptr = chunk_lens_ptr;

        inputs[0]->CopyFrom(void_chunk_lens_ptr);
        inputs[1]->CopyFrom(att_cache);
        inputs[2]->CopyFrom(cnn_cache);
        inputs[3]->CopyFrom(chunk_xs_ptr);
        inputs[4]->CopyFrom(cache_mask);
        inputs[5]->CopyFrom(offset);
        LOG_TS(m_ts, "wenet inference");
        auto status = net->Forward();
        LOG_TS(m_ts, "wenet inference");
        assert(BM_SUCCESS == status);
    
        outputs[0]->CopyTo(log_probs);
        outputs[1]->CopyTo(log_probs_idx);
        outputs[2]->CopyTo(chunk_out);
        outputs[3]->CopyTo(chunk_out_lens);
        outputs[4]->CopyTo(offset);
        outputs[5]->CopyTo(att_cache);
        outputs[6]->CopyTo(cnn_cache);
        outputs[7]->CopyTo(cache_mask);

        LOG_TS(m_ts, "wenet postprocess");
        std::vector<std::string> hyps = ctc_decoding(log_probs, log_probs_idx, chunk_out_lens, beam_size, batch_size, dict, "ctc_prefix_beam_search");
        LOG_TS(m_ts, "wenet postprocess");
        std::cout << hyps[0] << std::endl;
        result += hyps[0];

        std::free(chunk_lens_ptr);
        std::free(chunk_xs_ptr);
    }
    std::free(att_cache);
    std::free(cnn_cache);
    std::free(cache_mask);
    std::free(offset);

    std::free(log_probs);
    std::free(log_probs_idx);
    std::free(chunk_out);
    std::free(chunk_out_lens);
    
    return 0;
}

std::string WeNet::Recognize(const char* file_path) {
    LOG_TS(m_ts, "wenet preprocess");
    pre_process(file_path);
    LOG_TS(m_ts, "wenet preprocess");

    inference();
    return result;
}

void WeNet::enableProfile(TimeStamp *ts) {
    m_ts = ts;
}